akka-typed/src/main/scala/akka/typed/javadsl/BehaviorBuilder.scala

```diff
@@ -0,0 +1,141 @@
+/**
+ * Copyright (C) 2009-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.typed.javadsl
+
+import scala.annotation.tailrec
+import akka.japi.function.{ Function, Function2, Predicate }
+import akka.typed
+import akka.typed.{ Behavior, ExtensibleBehavior, Signal }
+import BehaviorBuilder._
+import akka.annotation.InternalApi
+
+/**
+ * Used for creating a [[akka.typed.Behavior]].
+ *
+ * There are methods for matching both messages and signals on type and (optionally) predicate.
+ *
+ * @tparam T the common superclass of all supported messages.
+ */
+case class BehaviorBuilder[T](
+  private val messageHandlers: List[Case[T, T]],
+  private val signalHandlers:  List[Case[T, Signal]]
+) extends ExtensibleBehavior[T] {
+
+  override def receiveMessage(ctx: typed.ActorContext[T], msg: T): Behavior[T] = receive[T](ctx, msg, messageHandlers)
+
+  override def receiveSignal(ctx: typed.ActorContext[T], msg: Signal): Behavior[T] = receive[Signal](ctx, msg, signalHandlers)
+
+  @tailrec
+  private def receive[M](ctx: ActorContext[T], msg: M, handlers: List[Case[T, M]]): Behavior[T] =
+    handlers match {
+      case Case(cls, predicate, handler) :: tail ⇒
+        if (cls.isAssignableFrom(msg.getClass) && (predicate.isEmpty || predicate.get.apply(msg))) handler(ctx, msg)
+        else receive[M](ctx, msg, tail)
+      case _ ⇒
+        // emulate scala match error
+        throw new MatchError(msg)
```

Comment:
We're pattern-matching on the list of cases here - but if none of the cases matches, we want to signal that the _message_ didn't match any of the cases. If we'd skip this case we'd get a `MatchError` for the empty list (as we exhausted all cases).